### PR TITLE
Fix box zoom

### LINKED ISSUE #1678: [BUG] Cubeviz: Box zoom does not include y axis bounds in selection
**Describe the bug**
Box zoom in the image viewers does not respect y axis scale.

https://user-images.githubusercontent.com/7179333/192575480-cea68a77-3517-4fc7-9d1b-31075f1ae3cb.mov

**To Reproduce**
1. Select box zoom tool
2. Create skinny rectangle subset from top of viewer to bottom
3. See unexpected behavior

**Expected behavior**
"The resulting axes limits may not exactly match the drawn region in order to preserve the equal-aspect ratio, but the entire drawn region should be included."

**Package versions (please complete the following information):**

**Additional context (e.g. data files)**

### PR DESCRIPTION
**Pull request by javerbukh**
### Description

This pull request is to address box zoom not respecting the aspect ratio of the viewer. You can see the associated issue for the original behavior, here is how it is in this PR:

https://user-images.githubusercontent.com/7179333/195392524-2db4eb6f-1bf4-4420-b53f-974510f3f147.mov

Fixes #1678 

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/core/tools.py [hunk 1]
```diff
@@ -144,8 +144,24 @@ def on_update_zoom(self):
...
+        new_interact_x = self.interact.selected_x
+
+        # If aspect is equal, then we need to preserve that aspect ratio
+        # even if the box zoom area does not.
+        if self.viewer.state.aspect == "equal":
+            aspect_ratio = ((self.viewer.state.x_max - self.viewer.state.x_min) /
+                            (self.viewer.state.y_max - self.viewer.state.y_min))
+
+            x_diff = self.interact.selected_x[1] - self.interact.selected_x[0]
+            y_diff = self.interact.selected_y[1] - self.interact.selected_y[0]
```
Comment (kecnry) [inline jdaviz/core/tools.py:156]: can `[1]` ever be smaller than `[0]` (inverted axes), and would that cause problems?  I suspect we wouldn't see that in jdaviz for any case that uses equal aspect ratio, so maybe just an inline comment stating if this assumption is made would suffice for now.

**Review by kecnry — COMMENTED** [review]
Seems to work as intended for me, nicely done!

File: jdaviz/core/tools.py (hunk 1 above)
Comment (javerbukh) [reply jdaviz/core/tools.py:156]: I tried to box zoom backwards and `[1]` is always the max value. I will add a comment to clarify that.

File: jdaviz/core/tools.py [hunk 2]
```diff
@@ -144,8 +144,24 @@ def on_update_zoom(self):
...
+
+            if x_diff < (y_diff / (1 / aspect_ratio)):
+                y_with_aspect = y_diff / (1 / aspect_ratio)
+                new_interact_x = (self.interact.selected_x[0] - y_with_aspect / 2,
+                                  self.interact.selected_x[1] + y_with_aspect / 2)
+
         self.viewer.state.y_min, self.viewer.state.y_max = self.interact.selected_y
+        self.viewer.state.x_min, self.viewer.state.x_max = new_interact_x
```
[the remark this answers is not in the document]
Comment (kecnry) [reply jdaviz/core/tools.py:164]: It seems to work as expected on my end, I just wonder if it _always_ will, or if we need logic to handle the opposite situation intead

File: jdaviz/core/tests/test_tools.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+    flux_viewer = cubeviz_helper.app.get_viewer('flux-viewer')
+
+    assert flux_viewer.state.y_min == -0.5
+    assert flux_viewer.state.y_max == 127.5
+    assert flux_viewer.state.x_min == -0.5
+    assert flux_viewer.state.x_max == 127.5
+
+    # t = flux_viewer.toolbar_nested.tools['jdaviz:boxzoom']
+    # t.activate()
```
[the remark this answers is not in the document]
Comment (javerbukh) [reply jdaviz/core/tests/test_tools.py]: I added something similar to what you said here, let me know if it works for you.

**Comment by javerbukh** [comment]
@camipacifici found that when using the box zoom tool in Imviz, the viewer axes adjust to the exact box zoom interactable when the x axis is much longer than the y. In the opposite case, the entire y range of values and x range of values are included in the freshly zoomed viewer, but there is some extra padding in the y axis that includes more values than were included in the box zoom. I'm wondering if this padding is being added upstream, do you know if that is the case @maartenbreddels or @astrofrog ?

File: jdaviz/core/tests/test_tools.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kecnry) [reply jdaviz/core/tests/test_tools.py]: looks reasonable to me!  I don't think these tools are considered public API yet... but if we ever do make them public, we might want to consider hiding this method (along with others - so I don't think you need to do it now).

File: jdaviz/core/tools.py [hunk 4]
```diff
@@ -144,9 +144,31 @@ def on_update_zoom(self):
...
+                              (self.viewer.state.y_max - self.viewer.state.y_min))
+
+            # The value at index of 1 will always be the larger of the two numbers
+            x_diff = self.interact.selected_x[1] - self.interact.selected_x[0]
+            y_diff = self.interact.selected_y[1] - self.interact.selected_y[0]
+
+            if x_diff < (y_diff / (1 / fig_axes_ratio)):
+                y_with_aspect = y_diff / (1 / fig_axes_ratio)
+                new_interact_x = (self.interact.selected_x[0] - y_with_aspect / 2,
+                                  self.interact.selected_x[1] + y_with_aspect / 2)
```
Comment (dhomeier) [inline jdaviz/core/tools.py:168]:
```suggestion
            if x_diff < y_diff * fig_axes_ratio:
                x_with_aspect = y_diff * fig_axes_ratio
                x_centre = (self.interact.selected_x[0] + self.interact.selected_x[1]) / 2
                new_interact_x = (x_centre - x_with_aspect / 2, x_centre + x_with_aspect / 2)
```
Squashed the multiple fractions here for clarity; but I think what you really want to resolve https://github.com/spacetelescope/jdaviz/pull/1726#issuecomment-1276552660 is to stretch the new x range symmetrically to scale with the (larger) y range.
What surprised however was that no such correction was required for `x_diff > y_diff * fig_axes_ratio` to begin with - `selected_y` seem already to be adjusted automatically. The reason behind this is apparently that
```
viewer.scales
{'x': LinearScale(allow_padding=False, max=2919.1, min=1384.2), 'y': LinearScale(max=1229.9, min=281.8)}
```
I don't know the reason for this asymmetry, but it seems you can get by without the entire post-correction above by setting instead
```python
    def _new_interact(self):
        x_scale = self.viewer.scale_x
        x_scale.allow_padding = True
        return BrushSelector(x_scale=scale_x, y_scale=self.viewer.scale_y, color=INTERACT_COLOR)
```

Comment (javerbukh) [reply jdaviz/core/tools.py]: Thank you for finding the cause and cleaning up the code! Yes we were wondering why the x axis seemed to work as expected but not the y. @camipacifici , I believe @dhomeier's changes fix the problem you were seeing before, at least it looks like that to me. Do you have time to check that?

File: jdaviz/core/tests/test_tools.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    cubeviz_helper.app.add_data_to_viewer('flux-viewer', 'Test Flux')
+
+    flux_viewer = cubeviz_helper.app.get_viewer('flux-viewer')
+
+    assert flux_viewer.state.y_min == -0.5
+    assert flux_viewer.state.y_max == 127.5
+    assert flux_viewer.state.x_min == -0.5
+    assert flux_viewer.state.x_max == 127.5
+
+    # Test does not work on python 3.8?
```
Comment (dhomeier) [inline jdaviz/core/tests/test_tools.py:19]: So it was not just my local installation (which is 3.10 on macOS)...

Comment (javerbukh) [reply jdaviz/core/tests/test_tools.py:19]: It worked on your installation or no?

Comment (dhomeier) [reply jdaviz/core/tests/test_tools.py:19]: It looks OK in the notebook, but the test fails with
```python
>       t = flux_viewer.toolbar_nested.tools['jdaviz:boxzoom']
E       AttributeError: 'CubevizImageView' object has no attribute 'toolbar_nested'
```

Comment (javerbukh) [reply jdaviz/core/tests/test_tools.py]: Good call! Waiting to see if tests pass now.

Comment (dhomeier) [reply jdaviz/core/tests/test_tools.py]: Works for me!

**Review by kecnry — APPROVED** [review]
Can't seem to break it and code definitely looks like code!  Thanks!

**Comment by camipacifici** [comment]
Works well for me, thank you!

**Comment by rosteen** [comment]
Rebasing now that #1799 is in should fix the 3.11 failure.

**Comment by pllim** [comment]
Still needs a change log.

**Review by pllim — APPROVED** [review]
Timeout error is unrelated.

I am approving by proxy since Cami basically approved.

Thanks!

**Pull request merged**